### PR TITLE
adds stream::enumerate combinator

### PR DESCRIPTION
**Pull request by montekki**
enumerate might be handy.
---
Stdlib: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate
Ref: #129

### CODE REVIEW

File: src/stream/stream/mod.rs [hunk 1]
```diff
@@ -136,6 +138,36 @@ pub trait Stream {
...
+    /// let mut s = stream::repeat(9).take(4).enumerate();
+    /// let mut c: usize = 0;
+    ///
+    /// while let Some((i, v)) = s.next().await {
+    ///     assert_eq!(c, i);
+    ///     assert_eq!(v, 9);
+    ///     c += 1;
+    /// }
+    /// #
+    /// # }) }
```
Comment (yoshuawuyts) [inline src/stream/stream/mod.rs]: This is a great test! But I feel like the example from `Iterator::enumerate` was a little clearer in its intent:
```rust
let a = ['a', 'b', 'c'];

let mut iter = a.iter().enumerate();

assert_eq!(iter.next(), Some((0, &'a')));
assert_eq!(iter.next(), Some((1, &'b')));
assert_eq!(iter.next(), Some((2, &'c')));
assert_eq!(iter.next(), None);
```

Perhaps manually calling `next().await` might make it easier to follow?

Comment (montekki) [reply src/stream/stream/mod.rs]: yep, makes sense

File: src/stream/stream/enumerate.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    pin_utils::unsafe_unpinned!(i: usize);
+
+    pub(super) fn new(stream: S) -> Self {
+        Enumerate { stream, i: 0 }
+    }
+}
+
+impl<S> futures_core::stream::Stream for Enumerate<S>
+where
+    S: Stream + Unpin + Sized,
```
Comment (stjepang) [inline src/stream/stream/enumerate.rs]:
```suggestion
    S: Stream,
```

**Pull request merged**